lib/hsa/mcwamp_hsa.cpp

```diff
@@ -1613,30 +1631,77 @@ class HSAQueue final : public KalmarQueue
             printAsyncOps(std::cerr);
         }
 
+        std::lock_guard<std::recursive_mutex> lg(qmutex);
 
-        {
-            std::lock_guard<std::recursive_mutex> lg(qmutex);
-            bool foundFirstValidOp = false;
-            for (int i = asyncOps.size()-1; i >= 0;  i--) {
-                if (asyncOps[i] != nullptr) {
-                    auto asyncOp = asyncOps[i];
-                    if (!foundFirstValidOp) {
-                        hsa_signal_t sig =  *(static_cast <hsa_signal_t*> (asyncOp->getNativeHandle()));
-                        assert(sig.handle != 0);
-                        foundFirstValidOp = true;
-                    }
-                    // wait on valid futures only
-                    std::shared_future<void>* future = asyncOp->getFuture();
-                    if (future && future->valid()) {
-                        future->wait();
-                    }
+        int first_nullptr = -1;
+        int lastWaitOp = asyncOps.size();
+        if (drainingQueue_) {
+            lastWaitOp = lastWaitOp * HCC_QUEUE_FLUSHING_RATIO / 100;
+        }
+
+        // Locate the first op that is not a nullptr, starting from the most recent op.
+        // The future->wait() might cause the op at the current index to become a nullptr,
+        // or it might cause older ops to become nullptrs. Once a nullptr is located, it is guaranteed
+        // that all ops older than the located nullptr will also be nulltpr, so we can stop the search.
+        // We may also not find a nullptr, or not find a valid op to wait upon.
+        for (int i = lastWaitOp-1; i >= 0;  i--) {
+            if (asyncOps[i] != nullptr) {
+                auto asyncOp = asyncOps[i];
+                // we only drain starting at first found marker
+                if (drainingQueue_ && asyncOp->getCommandKind() != hcCommandMarker) continue;
+                // wait on valid futures only
+                std::shared_future<void>* future = asyncOp->getFuture();
+                if (future && future->valid()) {
+                    future->wait();
```

Comment:
We are holding the lock to this asyncOps and wait so this may cause a potential deadlock here, need to take a closer look

Reply:
`future->wait()` happens in the original code for `HSAQueue::wait()`, so this deadlock might not be specific to my PR.   I'm looking at this code closely now, and indeed the thread holding the lock for `asyncOps` will trigger `hsa_signal_wait_scacquire` because of the `future->wait()`.  If queue stealing occurs by another thread while the signal wait is blocking, could this cause the deadlock?